Reproducible builds

petersalomonsen opened the pull request:
- all contracts building with `cargo near build reproducible-wasm`
- addressed comments from https://github.com/near-daos/sputnik-dao-contract/pull/218
  - notice my observation regarding adding `#[borsh(use_discriminant = true)]` when using the `near` attribute macro, which is also commented in the code ( you get the error `multiple `borsh` attributes not allowed`, should this be addressed in near-sdk? )
- minor clean-up
- update CI to use `cargo near` with reproducible builds
- MacOS test is removed due to docker not available

resolves #211

petersalomonsen commented:
@race-of-sloths include

race-of-sloths commented:
@petersalomonsen Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
We are launching a survey "Who is faster? You vs SpaceX"

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://race-of-sloths.com/images/bot-banner-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://race-of-sloths.com/images/bot-banner-light.svg">
    <img alt="Shows inviting banner with latest news." src="https://race-of-sloths.com/images/bot-banner-light.svg">
</picture>](https://race-of-sloths.com/latest-news)

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/petersalomonsen?type=bot&pr=near-daos/sputnik-dao-contract/220&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/petersalomonsen?type=bot&pr=near-daos/sputnik-dao-contract/220&theme=light">
    <img alt="Shows profile picture for the author of the PR" src="https://badge.race-of-sloths.com/petersalomonsen?type=bot&pr=near-daos/sputnik-dao-contract/220&theme=white">
</picture>](https://race-of-sloths.com/profile/petersalomonsen)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command. Alternatively, autoscoring [1,2] will be applied for this pull request

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/petersalomonsen)

For maintainers:
- Score pull requests that participate in the Race of Sloths and receive a reward
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

frol commented:
@petersalomonsen overall looks good to me! Actually, I remembered that serializers accept the parameters: https://github.com/near/near-sdk-rs/pull/1158, so you don’t need to use standalone #[serde()] or #[borsh()]

Please, update CI scripts as well

petersalomonsen commented:
> @petersalomonsen overall looks good to me! Actually, I remembered that serializers accept the parameters: [near/near-sdk-rs#1158](https://github.com/near/near-sdk-rs/pull/1158), so you don’t need to use standalone #[serde()] or #[borsh()]
> 
> Please, update CI scripts as well

Thanks for the pointer about the attribute macro @frol , and also for creating the [documentation ticket](https://github.com/near/near-sdk-rs/issues/1335).

The CI is also updated now to use the reproducible Wasm builds. I also enabled the release archive building on PR to be able to inspect/preview the release archive.

Note that I have removed the MacOS test since `docker` is not available. Might be possible to use e.g. `colima`, but I'm not sure if the MacOS test is crucial here? Do we want to keep it?

frol reviewed (approved):
@race-of-sloths score 5
